File: cpp/mpi/src/mpi_case_driver.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#ifndef PATH_MAX
#define PATH_MAX 4096
#endif

typedef struct {
    int N;
    int Re;
    const char *scheme;
    const char *pressure;
    const char *implementation;
} CaseDef;

static void mkdir_p(const char *path) {
    char tmp[PATH_MAX];
    snprintf(tmp, sizeof(tmp), "%s", path);
    for (char *p = tmp + 1; *p; ++p) {
        if (*p == '/') { *p = 0; mkdir(tmp, 0775); *p = '/'; }
    }
    mkdir(tmp, 0775);
}

static int build_cases(const char *mode, CaseDef *cases, int max_cases) {
    int meshes[3] = {32, 64, 128}; int n_mesh = 3;
    int res[3] = {100, 400, 1000}; int n_re = 3;
    const char *schemes[2] = {"upwind", "central"}; int n_scheme = 2;
    const char *pressures[2] = {"RBGS", "RBSOR"}; int n_pressure = 2;
    const char *impls[1] = {"serial_cpp"}; int n_impl = 1;

    if (strcmp(mode, "smoke") == 0) { meshes[0] = 16; n_mesh = 1; res[0] = 100; n_re = 1; n_scheme = 1; n_pressure = 1; }
    else if (strcmp(mode, "quick") == 0) { n_mesh = 2; n_re = 2; }
    else if (strcmp(mode, "medium") == 0) { n_mesh = 2; n_re = 3; }
    else if (strcmp(mode, "full") == 0) { }
    else { fprintf(stderr, "Unknown mode: %s\n", mode); MPI_Abort(MPI_COMM_WORLD, 2); }

    int n = 0;
    for (int a = 0; a < n_mesh; ++a)
    for (int b = 0; b < n_re; ++b)
    for (int c = 0; c < n_scheme; ++c)
    for (int d = 0; d < n_pressure; ++d)
    for (int e = 0; e < n_impl; ++e) {
        if (n >= max_cases) return n;
        cases[n++] = (CaseDef){meshes[a], res[b], schemes[c], pressures[d], impls[e]};
    }
    return n;
}

int main(int argc, char **argv) {
    MPI_Init(&argc, &argv);
    int rank = 0, size = 1;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    const char *mode = "quick";
    const char *solver = "./bin/lid_cavity";
    for (int i = 1; i < argc; ++i) {
        if (strcmp(argv[i], "--mode") == 0 && i + 1 < argc) mode = argv[++i];
        else if (strcmp(argv[i], "--solver") == 0 && i + 1 < argc) solver = argv[++i];
    }

    char root[PATH_MAX];
    if (!getcwd(root, sizeof(root))) { perror("getcwd"); MPI_Abort(MPI_COMM_WORLD, 3); }
    char solver_abs[PATH_MAX];
    if (solver[0] == '/') snprintf(solver_abs, sizeof(solver_abs), "%s", solver);
    else snprintf(solver_abs, sizeof(solver_abs), "%s/%s", root, solver);

    CaseDef cases[512];
    int ncases = build_cases(mode, cases, 512);
    if (rank == 0) printf("MPI case-parallel C++ driver: %d cases over %d ranks\n", ncases, size);

    for (int case_id = rank; case_id < ncases; case_id += size) {
        CaseDef c = cases[case_id];
        char workdir[PATH_MAX];
        snprintf(workdir, sizeof(workdir), "%s/results/mpi_raw/rank_%03d/case_%03d", root, rank, case_id + 1);
        mkdir_p(workdir);
        char cmd[PATH_MAX * 2];
        snprintf(cmd, sizeof(cmd),
                 "cd '%s' && '%s' --single --N %d --Re %d --scheme %s --pressure %s --implementation %s --no-fields",
                 workdir, solver_abs, c.N, c.Re, c.scheme, c.pressure, c.implementation);
        printf("[rank %d] case %03d: N=%d Re=%d %s %s %s\n", rank, case_id + 1, c.N, c.Re, c.scheme, c.pressure, c.implementation);
        int rc = system(cmd);
        if (rc != 0) fprintf(stderr, "[rank %d] command failed for case %d with rc=%d\n", rank, case_id + 1, rc);
    }

    MPI_Barrier(MPI_COMM_WORLD);
    if (rank == 0) printf("Done. Run: python3 scripts/merge_mpi_results.py\n");
    MPI_Finalize();
    return 0;
}

File: python/mpi/src/mpi_case_driver.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import csv
from pathlib import Path
import sys

from mpi4py import MPI

sys.path.insert(0, str(Path(__file__).resolve().parent))

from lidcavity.config import Config
from lidcavity.cli import configure_mode
from lidcavity.solver import solve_lid_cavity
from lidcavity.validation import validate_against_ghia, quality_label
from lidcavity.io_utils import summary_header, summary_row, write_history_csv, write_field_csv
from lidcavity.utils import lower, upper, normalize_implementation


def build_cases(mode: str):
    cfg = Config()
    configure_mode(cfg, mode)

    cases = []
    case_id = 0

    for N in cfg.meshes:
        for Re in cfg.re_list:
            for scheme in cfg.schemes:
                for pressure in cfg.pressure_solvers:
                    for impl in cfg.implementations:
                        case_id += 1
                        cases.append((case_id, N, Re, scheme, pressure, impl))

    return cases


def append_row(path: Path, header, row):
    new_file = not path.exists()
    path.parent.mkdir(parents=True, exist_ok=True)

    with path.open("a", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        if new_file:
            w.writerow(header)
        w.writerow(row)


def merge_rank_files(part: str):
    out_dir = Path("results/data")
    out_dir.mkdir(parents=True, exist_ok=True)

    rank_files = sorted(out_dir.glob(f"study_summary_mpi_part_{part}_rank_*.csv"))
    out_path = out_dir / f"study_summary_mpi_part_{part}.csv"

    rows = []
    header = None

    for rf in rank_files:
        with rf.open("r", newline="", encoding="utf-8") as f:
            reader = csv.reader(f)
            file_header = next(reader, None)
            if file_header and header is None:
                header = file_header
            for row in reader:
                if row:
                    rows.append(row)

    if header is None:
        header = summary_header()

    # Deduplicate by case id, keep last
    by_case = {}
    for row in rows:
        by_case[int(row[0])] = row

    rows = [by_case[k] for k in sorted(by_case)]

    with out_path.open("w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(header)
        w.writerows(rows)

    print(f"Wrote {out_path}")
    print(f"Rows written: {len(rows)}")


def main() -> int:
    parser = argparse.ArgumentParser(description="Safe checkpointed MPI case-parallel Python lid-cavity runner")
    parser.add_argument("--mode", choices=["smoke", "quick", "medium", "full"], default="full")
    parser.add_argument("--no-fields", action="store_true")
    parser.add_argument("--case-start", type=int, default=1)
    parser.add_argument("--case-end", type=int, default=None)
    parser.add_argument("--part", type=str, default="000")
    args = parser.parse_args()

    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    all_cases = build_cases(args.mode)
    total_cases = len(all_cases)

    case_end = args.case_end if args.case_end is not None else total_cases
    selected = [c for c in all_cases if args.case_start <= c[0] <= case_end]

    if rank == 0:
        print("Python MPI checkpointed driver")
        print(f"Total cases in mode {args.mode}: {total_cases}")
        print(f"Selected cases: {args.case_start} to {case_end}")
        print(f"Selected count: {len(selected)}")
        print(f"MPI ranks: {size}", flush=True)

    local_path = Path("results/data") / f"study_summary_mpi_part_{args.part}_rank_{rank:03d}.csv"

    for local_index in range(rank, len(selected), size):
        case_id, N, Re, scheme, pressure, impl = selected[local_index]

        cfg = Config()
        configure_mode(cfg, "single")
        cfg.meshes = [N]
        cfg.re_list = [Re]
        cfg.schemes = [lower(scheme)]
        cfg.pressure_solvers = [upper(pressure)]
        cfg.implementations = [normalize_implementation(impl)]
        cfg.save_fields = not args.no_fields

        cfg.results_dir = f"results/mpi_raw/part_{args.part}/rank_{rank:03d}/case_{case_id:03d}/results"
        cfg.data_dir = str(Path(cfg.results_dir) / "data")
        Path(cfg.data_dir).mkdir(parents=True, exist_ok=True)

        print(
            f"[rank {rank}] case {case_id:03d}: "
            f"N={N} Re={Re} scheme={scheme} pressure={pressure} impl={impl}",
            flush=True,
        )

        r = solve_lid_cavity(N, Re, scheme, pressure, impl, cfg)
        m = validate_against_ghia(r, cfg)
        q = quality_label(r, m)

        original_impl = r.implementation
        r.implementation = "mpi_python_case_parallel_" + original_impl.replace("serial_python_", "")

        case_name = f"case_{case_id:03d}_N{N}_Re{Re}_{lower(scheme)}_{upper(pressure)}_{lower(r.implementation)}"

        write_history_csv(r, case_name, cfg)
        if cfg.save_fields:
            write_field_csv(r, case_name, cfg)

        row = summary_row(case_id, r, m, q)
        append_row(local_path, summary_header(), row)

        print(f"[rank {rank}] finished case {case_id:03d} and checkpointed row", flush=True)

    comm.Barrier()

    if rank == 0:
        merge_rank_files(args.part)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
